File: lib/mathlib/mathlib.go
package mathlib

import (
	"math"
	"math/rand"
	"time"

	"github.com/arnodel/golua/lib/packagelib"
	rt "github.com/arnodel/golua/runtime"
)

var LibLoader = packagelib.Loader{
	Load: load,
	Name: "math",
}

func load(r *rt.Runtime) (rt.Value, func()) {
	pkg := rt.NewTable()
	r.SetEnv(pkg, "huge", rt.FloatValue(math.Inf(1)))
	r.SetEnv(pkg, "maxinteger", rt.IntValue(math.MaxInt64))
	r.SetEnv(pkg, "mininteger", rt.IntValue(math.MinInt64))
	r.SetEnv(pkg, "pi", rt.FloatValue(math.Pi))

	rt.SolemnlyDeclareCompliance(
		rt.ComplyCpuSafe|rt.ComplyMemSafe|rt.ComplyTimeSafe|rt.ComplyIoSafe,

		r.SetEnvGoFunc(pkg, "abs", abs, 1, false),
		r.SetEnvGoFunc(pkg, "acos", acos, 1, false),
		r.SetEnvGoFunc(pkg, "asin", asin, 1, false),
		r.SetEnvGoFunc(pkg, "atan", atan, 2, false),
		r.SetEnvGoFunc(pkg, "ceil", ceil, 1, false),
		r.SetEnvGoFunc(pkg, "cos", cos, 1, false),
		r.SetEnvGoFunc(pkg, "deg", deg, 1, false),
		r.SetEnvGoFunc(pkg, "exp", exp, 1, false),
		r.SetEnvGoFunc(pkg, "floor", floor, 1, false),
		r.SetEnvGoFunc(pkg, "fmod", fmod, 2, false),
		r.SetEnvGoFunc(pkg, "log", log, 2, false),
		r.SetEnvGoFunc(pkg, "max", max, 1, true),
		r.SetEnvGoFunc(pkg, "min", min, 1, true),
		r.SetEnvGoFunc(pkg, "modf", modf, 1, false),
		r.SetEnvGoFunc(pkg, "rad", rad, 1, false),
		r.SetEnvGoFunc(pkg, "random", random, 2, false),
		r.SetEnvGoFunc(pkg, "randomseed", randomseed, 2, false),
		r.SetEnvGoFunc(pkg, "sin", sin, 1, false),
		r.SetEnvGoFunc(pkg, "sqrt", sqrt, 1, false),
		r.SetEnvGoFunc(pkg, "tan", tan, 1, false),
		r.SetEnvGoFunc(pkg, "tointeger", tointeger, 1, false),
		r.SetEnvGoFunc(pkg, "type", typef, 1, false),
		r.SetEnvGoFunc(pkg, "ult", ult, 2, false),
	)

	return rt.TableValue(pkg), nil
}

func abs(t *rt.Thread, c *rt.GoCont) (rt.Cont, *rt.Error) {
	if err := c.Check1Arg(); err != nil {
		return nil, err
	}
	next := c.Next()
	n, f, tp := rt.ToNumber(c.Arg(0))
	switch tp {
	case rt.IsInt:
		if n < 0 {
			n = -n
		}
		t.Push1(next, rt.IntValue(n))
	case rt.IsFloat:
		t.Push1(next, rt.FloatValue(math.Abs(f)))
	default:
		return nil, rt.NewErrorS("#1 must be a number")
	}
	return next, nil
}

func acos(t *rt.Thread, c *rt.GoCont) (rt.Cont, *rt.Error) {
	if err := c.Check1Arg(); err != nil {
		return nil, err
	}
	x, err := c.FloatArg(0)
	if err != nil {
		return nil, err
	}
	y := rt.FloatValue(math.Acos(x))
	return c.PushingNext1(t.Runtime, y), nil
}

func asin(t *rt.Thread, c *rt.GoCont) (rt.Cont, *rt.Error) {
	if err := c.Check1Arg(); err != nil {
		return nil, err
	}
	x, err := c.FloatArg(0)
	if err != nil {
		return nil, err
	}
	y := rt.FloatValue(math.Asin(x))
	return c.PushingNext1(t.Runtime, y), nil
}

func atan(t *rt.Thread, c *rt.GoCont) (rt.Cont, *rt.Error) {
	if err := c.Check1Arg(); err != nil {
		return nil, err
	}
	y, err := c.FloatArg(0)
	if err != nil {
		return nil, err
	}
	var x float64 = 1
	if c.NArgs() >= 2 {
		x, err = c.FloatArg(1)
		if err != nil {
			return nil, err
		}
	}
	z := rt.FloatValue(math.Atan2(y, x))
	return c.PushingNext1(t.Runtime, z), nil
}

func ceil(t *rt.Thread, c *rt.GoCont) (rt.Cont, *rt.Error) {
	if err := c.Check1Arg(); err != nil {
		return nil, err
	}
	next := c.Next()
	n, f, tp := rt.ToNumber(c.Arg(0))
	switch tp {
	case rt.IsInt:
		t.Push1(next, rt.IntValue(n))
	case rt.IsFloat:
		f = math.Ceil(f)
		n = int64(f)
		if float64(n) == f {
			t.Push1(next, rt.IntValue(n))
		} else {
			t.Push1(next, rt.FloatValue(f))
		}
	default:
		return nil, rt.NewErrorS("#1 must be a number")
	}
	return next, nil
}

func cos(t *rt.Thread, c *rt.GoCont) (rt.Cont, *rt.Error) {
	if err := c.Check1Arg(); err != nil {
		return nil, err
	}
	x, err := c.FloatArg(0)
	if err != nil {
		return nil, err
	}
	y := rt.FloatValue(math.Cos(x))
	return c.PushingNext1(t.Runtime, y), nil
}

func deg(t *rt.Thread, c *rt.GoCont) (rt.Cont, *rt.Error) {
	if err := c.Check1Arg(); err != nil {
		return nil, err
	}
	x, err := c.FloatArg(0)
	if err != nil {
		return nil, err
	}
	y := rt.FloatValue(x * 180 / math.Pi)
	return c.PushingNext1(t.Runtime, y), nil
}

func exp(t *rt.Thread, c *rt.GoCont) (rt.Cont, *rt.Error) {
	if err := c.Check1Arg(); err != nil {
		return nil, err
	}
	x, err := c.FloatArg(0)
	if err != nil {
		return nil, err
	}
	y := rt.FloatValue(math.Exp(x))
	return c.PushingNext1(t.Runtime, y), nil
}

func floor(t *rt.Thread, c *rt.GoCont) (rt.Cont, *rt.Error) {
	if err := c.Check1Arg(); err != nil {
		return nil, err
	}
	next := c.Next()
	n, f, tp := rt.ToNumber(c.Arg(0))
	switch tp {
	case rt.IsInt:
		t.Push1(next, rt.IntValue(n))
	case rt.IsFloat:
		f = math.Floor(f)
		n = int64(f)
		if float64(n) == f {
			t.Push1(next, rt.IntValue(n))
		} else {
			t.Push1(next, rt.FloatValue(f))
		}
	default:
		return nil, rt.NewErrorS("#1 must be a number")
	}
	return next, nil
}

func fmod(t *rt.Thread, c *rt.GoCont) (rt.Cont, *rt.Error) {
	if err := c.CheckNArgs(2); err != nil {
		return nil, err
	}
	x, _ := rt.ToNumberValue(c.Arg(0))
	y, _ := rt.ToNumberValue(c.Arg(1))
	res, ok, err := rt.Mod(x, y)
	if !ok {
		err = rt.NewErrorS("expected numeric arguments")
	}
	if err != nil {
		return nil, err
	}
	return c.PushingNext1(t.Runtime, res), nil
}

func log(t *rt.Thread, c *rt.GoCont) (rt.Cont, *rt.Error) {
	if err := c.Check1Arg(); err != nil {
		return nil, err
	}
	x, err := c.FloatArg(0)
	if err != nil {
		return nil, err
	}
	y := math.Log(x)
	if c.NArgs() >= 2 {
		b, err := c.FloatArg(1)
		if err != nil {
			return nil, err
		}
		y = y / math.Log(b)
	}
	return c.PushingNext1(t.Runtime, rt.FloatValue(y)), nil
}

func max(t *rt.Thread, c *rt.GoCont) (rt.Cont, *rt.Error) {
	if err := c.Check1Arg(); err != nil {
		return nil, err
	}
	x := c.Arg(0)
	for _, y := range c.Etc() {
		lt, err := rt.Lt(t, x, y)
		if err != nil {
			return nil, err
		}
		if lt {
			x = y
		}
	}
	return c.PushingNext1(t.Runtime, x), nil
}

func min(t *rt.Thread, c *rt.GoCont) (rt.Cont, *rt.Error) {
	if err := c.Check1Arg(); err != nil {
		return nil, err
	}
	x := c.Arg(0)
	for _, y := range c.Etc() {
		lt, err := rt.Lt(t, y, x)
		if err != nil {
			return nil, err
		}
		if lt {
			x = y
		}
	}
	return c.PushingNext1(t.Runtime, x), nil
}

func modf(t *rt.Thread, c *rt.GoCont) (rt.Cont, *rt.Error) {
	if err := c.Check1Arg(); err != nil {
		return nil, err
	}
	next := c.Next()
	arg := c.Arg(0)
	if _, ok := arg.TryInt(); ok {
		t.Push1(next, arg)
		t.Push1(next, rt.FloatValue(0))
		return next, nil
	}

	x, ok := arg.TryFloat()
	if !ok {
		return nil, rt.NewErrorS("#1 must be numeric")
	}
	var i, f float64
	if math.IsInf(x, 0) {
		i, f = x, 0
	} else {
		i, f = math.Modf(x)
	}
	t.Push1(next, rt.FloatValue(i))
	t.Push1(next, rt.FloatValue(f))
	return next, nil
}

func rad(t *rt.Thread, c *rt.GoCont) (rt.Cont, *rt.Error) {
	if err := c.Check1Arg(); err != nil {
		return nil, err
	}
	x, err := c.FloatArg(0)
	if err != nil {
		return nil, err
	}
	y := rt.FloatValue(x * math.Pi / 180)
	return c.PushingNext1(t.Runtime, y), nil
}

// TODO: have a per runtime random generator
func random(t *rt.Thread, c *rt.GoCont) (rt.Cont, *rt.Error) {
	var (
		err *rt.Error
		m   int64 = 1
		n   int64
	)
	switch c.NArgs() {
	case 0:
		return c.PushingNext1(t.Runtime, rt.FloatValue(rand.Float64())), nil
	case 1:
		n, err = c.IntArg(0)
	case 2:
		m, err = c.IntArg(0)
		if err == nil {
			n, err = c.IntArg(1)
		}
	}
	if err != nil {
		return nil, err
	}
	if m > n {
		return nil, rt.NewErrorS("#2 must be >= #1")
	}
	var r int64
	if m <= 0 && m+math.MaxInt64 < n {
		return nil, rt.NewErrorS("interval too large")
	} else if m+math.MaxInt64 == n {
		r = rand.Int63()
	} else {
		r = rand.Int63n(n - m + 1)
	}
	return c.PushingNext1(t.Runtime, rt.IntValue(m+r)), nil
}

func randomseed(t *rt.Thread, c *rt.GoCont) (rt.Cont, *rt.Error) {
	var seed int64
	var err *rt.Error
	switch c.NArgs() {
	case 0:
		// Something "random"
		seed = time.Now().UnixNano()
	case 1:
		seed, err = c.IntArg(0)
		if err != nil {
			return nil, err
		}
	case 2:
		seed, err = c.IntArg(0)
		if err != nil {
			return nil, err
		}
		seed2, err := c.IntArg(1)
		if err != nil {
			return nil, err
		}
		// In Go the seed is only 64 bits so we mangle the seeds
		seed ^= seed2
	}
	rand.Seed(seed)
	return c.PushingNext(t.Runtime, rt.IntValue(seed), rt.IntValue(0)), nil
}

func sin(t *rt.Thread, c *rt.GoCont) (rt.Cont, *rt.Error) {
	if err := c.Check1Arg(); err != nil {
		return nil, err
	}
	x, err := c.FloatArg(0)
	if err != nil {
		return nil, err
	}
	y := rt.FloatValue(math.Sin(x))
	return c.PushingNext1(t.Runtime, y), nil
}

func sqrt(t *rt.Thread, c *rt.GoCont) (rt.Cont, *rt.Error) {
	if err := c.Check1Arg(); err != nil {
		return nil, err
	}
	x, err := c.FloatArg(0)
	if err != nil {
		return nil, err
	}
	y := rt.FloatValue(math.Sqrt(x))
	return c.PushingNext1(t.Runtime, y), nil
}

func tan(t *rt.Thread, c *rt.GoCont) (rt.Cont, *rt.Error) {
	if err := c.Check1Arg(); err != nil {
		return nil, err
	}
	x, err := c.FloatArg(0)
	if err != nil {
		return nil, err
	}
	y := rt.FloatValue(math.Tan(x))
	return c.PushingNext1(t.Runtime, y), nil
}

func tointeger(t *rt.Thread, c *rt.GoCont) (rt.Cont, *rt.Error) {
	if err := c.Check1Arg(); err != nil {
		return nil, err
	}
	n, err := c.IntArg(0)
	if err != nil {
		return c.PushingNext1(t.Runtime, rt.NilValue), nil
	}
	return c.PushingNext1(t.Runtime, rt.IntValue(n)), nil
}

func typef(t *rt.Thread, c *rt.GoCont) (rt.Cont, *rt.Error) {
	if err := c.Check1Arg(); err != nil {
		return nil, err
	}
	var tp rt.Value
	switch c.Arg(0).NumberType() {
	case rt.IntType:
		tp = rt.StringValue("integer")
	case rt.FloatType:
		tp = rt.StringValue("float")
	}
	return c.PushingNext1(t.Runtime, tp), nil
}

func ult(t *rt.Thread, c *rt.GoCont) (rt.Cont, *rt.Error) {
	if err := c.CheckNArgs(2); err != nil {
		return nil, err
	}
	x, err := c.IntArg(0)
	var y int64
	if err == nil {
		y, err = c.IntArg(1)
	}
	if err != nil {
		return nil, err
	}
	lt := rt.BoolValue(uint64(x) < uint64(y))
	return c.PushingNext1(t.Runtime, lt), nil
}

File: lib/mathlib/lua/mathlib.lua
local function checknumarg(f)
    local ok = not pcall(f) and not pcall(f, {})
    if ok then
        print "ok"
    else
        print "bad"
    end
end

do
    checknumarg(math.abs)
    --> =ok

    print(math.abs(12))
    --> =12

    print(math.abs(-12))
    --> =12

    print(math.abs(-0.5))
    --> =0.5

    print(math.abs("-123"))
    --> =123

    print(pcall(math.abs))
    --> ~false\t.*value needed

    print(pcall(math.abs, {}))
    --> ~false\t.*must be a number
end

do
    checknumarg(math.ceil)
    --> =ok

    print(math.ceil(123))
    --> =123

    print(math.ceil(12.7))
    --> =13

    print(math.ceil(-8.2))
    --> =-8

    print(math.ceil("888.5673"))
    --> =889
end

do
    checknumarg(math.exp)
    --> =ok

    print(math.exp(0))
    --> =1
    
end

do
    checknumarg(math.floor)
    --> =ok

    print(math.floor(123))
    --> =123

    print(math.floor(12.7))
    --> =12

    print(math.floor(-8.2))
    --> =-9

    print(math.floor("888.5673"))
    --> =888
end

do
    checknumarg(math.log)
    --> =ok

    print(math.log(1))
    --> =0
end

do
    print(math.max(3, 5, 2, 1))
    --> =5

    print(math.min(3, 5, 2, 1))
    --> =1
end

do
    checknumarg(math.modf)
    --> =ok

    print(math.modf(1.5))
    --> =1	0.5

    print(math.modf(-1.5))
    --> =-1	-0.5

    print(math.modf(1/0))
    --> =+Inf	0

    print(math.modf(-1/0))
    --> =-Inf	0

end

do
    checknumarg(math.rad)
    --> =ok

    checknumarg(math.deg)
    --> =ok

    print(math.rad(90) == math.pi / 2)
    --> =true

    print(math.deg(math.pi) == 180)
    --> =true
end

do
    checknumarg(math.randomseed)
    --> =ok

    local r = math.random()
    print(r >= 0 and r <= 1)
    --> =true

    local rands = {}
    math.randomseed(5)
    for i = 1, 20 do
        rands[i] = math.random(i, 2*i)
    end

    math.randomseed(5)
    local out, diff = 0, 0
    for i = 1, 20 do
        r = math.random(i, 2*i)
        if r < i or r > 2*i then
            out = out + 1
        end
        if rands[i] ~= r then
            diff = diff + 1
        end
    end
    print(out, diff)
    --> =0	0
end

do
    checknumarg(math.sin)
    --> =ok

    checknumarg(math.cos)
    --> =ok

    checknumarg(math.tan)
    --> =ok

    local function toz(f)
        return function(x)
            local y = f(x)
            if math.abs(y) < 1e-15 then
                return 0
            elseif math.abs(y) > 1e15 then
                return 1/0
            end
            return y
        end
    end
    local pi = math.pi
    local sin = toz(math.sin)
    local cos = toz(math.cos)
    local tan = toz(math.tan)

    print(sin(0), sin(pi/2), sin(pi), sin(3*pi/2))
    --> =0	1	0	-1

    print(cos(0), cos(pi/2), cos(pi), cos(3*pi/2))
    --> =1	0	-1	0

    print(tan(0), tan(pi/2), tan(pi), tan(3*pi/2))
    --> =0	+Inf	0	+Inf
end

do
    checknumarg(math.asin)
    --> =ok

    checknumarg(math.acos)
    --> =ok

    checknumarg(math.atan)
    --> =ok

    local function round(num) 
        if num >= 0 then return math.floor(num+.5) 
        else return math.ceil(num-.5) end
    end

    local function top(f)
        return function(x)
            local y = 4*f(x)/math.pi
            local ry = round(y)
            if math.abs(y - ry) > 1e-15 then
                return ry
            end
            return y
        end
    end

    local acos = top(math.acos)
    local asin = top(math.asin)
    local atan = top(math.atan)

    print(acos(-1), acos(0), acos(1))
    --> =4	2	0

    print(asin(-1), asin(0), asin(1))
    --> =-2	0	2

    print(atan(-1), atan(0), atan(1))
    --> =-1	0	1
end

do
    checknumarg(math.sqrt)
    --> =ok

    for i = 1, 5 do
        print(math.floor(1000*math.sqrt(i)))
    end
    --> =1000
    --> =1414
    --> =1732
    --> =2000
    --> =2236
end

do
    print(math.tointeger(56.0))
    --> =56

    print(math.tointeger("-123"))
    --> =-123

    print(math.tointeger({}))
    --> =nil

    print(math.tointeger(true))
    --> =nil
end

do
    print(math.type("123"))
    --> =nil

    print(math.type(123))
    --> =integer

    print(math.type(123.0))
    --> =float

    print(math.type(nil))
    --> =nil
end

do
    print(math.ult(1, 2))
    --> =true

    print(math.ult(-1, -2))
    --> =false

    print(math.ult(-1, 2))
    --> =false
end

do
    print(math.fmod(5, 2))
    --> =1

    -- TODO: fix implementation
    -- print(math.fmod(-5, 2))
    -- --> =-1
end
